### ISSUE TITLE
Build from source misses openssl/rand.h

### ISSUE THREAD

**Issue by hannesa2** [sqlcipher/android-database-sqlcipher#447]
I try from latest master cec41b9e7ef66adfb to build from source. 
I followed README.md instruction with
```
make init
export ANDROID_NDK_ROOT=~/Development/android-ndk-r15c/
make build-debug
```
and run into 
> fatal error: openssl/rand.h: No such file or directory

<img width="1229" alt="image" src="https://user-images.githubusercontent.com/3314607/59182032-f7b1e000-8b68-11e9-8224-55e5840c6629.png">

Where must the openssl located ?
Please update the build documentation to be able to build from source. Thank you

**Comment by developernotes** [comment]
Hi @hannesa2 

Try running `make distclean`, then `make build-debug` to see if that changes the behavior on your machine.

**Comment by hannesa2** [comment]
<img width="1290" alt="image" src="https://user-images.githubusercontent.com/3314607/59288803-e5bd6380-8c74-11e9-8360-1a76a2a20ec3.png">

**Comment by developernotes** [comment]
Hi @hannesa2 

Do you have any local modifications within the repository?

**Comment by developernotes** [comment]
@hannesa2 

What does the following report on your system: ```gcc --version```

**Comment by hannesa2** [comment]
There are no local changes, that's why I mention SHA1

<img width="1413" alt="image" src="https://user-images.githubusercontent.com/3314607/59304254-25e20d80-8c98-11e9-8a30-69c37686bc18.png">

**Comment by developernotes** [comment]
Hi @hannesa2 

I am not able to reproduce this issue. Would you try modifying the `net_sqlcipher_CursorWindow.cpp` file located in `android-database-sqlcipher/android-database-sqlcipher/src/main/cpp` directory and adjusting near the top from this:

```
#include <stdint.h>
#include <inttypes.h>
```

to this:

```
#include <stdint.h>
#define __STDC_FORMAT_MACROS
#include <inttypes.h>
```

Would you try that and let us know if this resolves the issue? Thanks!

**Comment by hannesa2** [comment]
As a maintainer I would love when I could point to Travis (or similar) CI job, where I could show as evidence that our repo builds properly.
Because you can't, I have my doubt what you deliver or how do your build this.
This is not only complain about this, **I want to help to improve this !** 

Btw I followed your advice, and here is the output 
https://github.com/sqlcipher/android-database-sqlcipher/pull/452

But anyway I want to say: **thank you**, now the build makes a step further to success, but next issue has to be solved 

> threads_pthread.c:(.text+0x18): undefined reference to `pthread_atfork'

**Comment by developernotes** [comment]
Hi @hannesa2 

Would you try deleting your local Android NDK and downloading a fresh copy of r15c to rule out any strange local corruption issues?

**Comment by hannesa2** [comment]
You are very fast in closing PR's and stop discussion on it. I opened https://github.com/sqlcipher/android-database-sqlcipher/pull/452 to show **exactly** what I did.

A new NDK makes no diff, can **you** @developernotes  run this successful on your computer ? 
I'm not.

```
mkdir ~/CipherbuildTest
cd ~/CipherbuildTest
wget https://dl.google.com/android/repository/android-ndk-r15c-darwin-x86_64.zip
unzip android-ndk-r15c-darwin-x86_64.zip 
export ANDROID_NDK_ROOT=~/CipherbuildTest/android-ndk-r15c/
git clone --depth 1 https://github.com/sqlcipher/android-database-sqlcipher.git 
cd android-database-sqlcipher/
make init
make build-debug
```

**Comment by developernotes** [comment]
Hi @hannesa2 

>You are very fast in closing PR's and stop discussion on it. I opened #452 to show exactly what I did.

No malice was intended there, we experience random "fly by" comments on tickets quite often that are closed. Unfortunately, GitHub Issues does not offer the appropriate level of control on tickets. We would certainly rather close a ticket, but still allow the original reporter and repo owners _only_ to continue, but GitHub does not currently support that. Unfortunately, the #452 only shows the next error you received, but your sample above is helpful.

```mkdir ~/CipherbuildTest
cd ~/CipherbuildTest
wget https://dl.google.com/android/repository/android-ndk-r15c-darwin-x86_64.zip
unzip android-ndk-r15c-darwin-x86_64.zip 
export ANDROID_NDK_ROOT=~/CipherbuildTest/android-ndk-r15c/
git clone --depth 1 https://github.com/sqlcipher/android-database-sqlcipher.git 
cd android-database-sqlcipher/
make init
make build-debug
```

I was able to get this to work on my machine, however performing the exact steps you define above produces an error when attempting to build OpenSSL:

```
no NDK arm-linux-androideabi-gcc on $PATH at (eval 5) line 124.
```

I resolved that by setting `ANDROID_NDK_HOME` to the path of `ANDROID_NDK_ROOT`. Would you try your steps above, but adding this line (if you are using bash or the equivalent) and report back?

```
export ANDROID_NDK_HOME=$ANDROID_NDK_ROOT
```

**Comment by hannesa2** [comment]
You made it work ? I'm not. I switched to the branch with your suggested fix 

```
mkdir ~/CipherbuildTest
cd ~/CipherbuildTest
wget https://dl.google.com/android/repository/android-ndk-r15c-darwin-x86_64.zip
unzip android-ndk-r15c-darwin-x86_64.zip 
export ANDROID_NDK_ROOT=~/CipherbuildTest/android-ndk-r15c/
export ANDROID_NDK_HOME=$ANDROID_NDK_ROOT
git clone -b FollowAdviceByNickParker --depth 1 https://github.com/hannesa2/android-database-sqlcipher.git
cd android-database-sqlcipher/
make init
make build-debug
```

and still run into 
![image](https://user-images.githubusercontent.com/3314607/59744545-14d66500-9273-11e9-9416-1cb3588a2a0b.png)


Btw, this all is a task of a CI

**Comment by sjlombardo** [comment]
Hello @hannesa2 - If I might ask, why did you switch the branch to your custom changes, instead of just checking out master as specified in the build steps that you originally suggested? The changes that you made on `FollowAdviceByNickParker` were only experimental, simply to try working around your specific reported problems. Now that we have identified an environmental issue (i.e. that your system was missing a specific environmental variable required by the NDK), there is no reason to use the custom branch, particularly given that @developernotes verified your build steps were working on master.

As an aside, posting a screenshot of a generic error message from gradle is not particularly helpful for troubleshooting purposes. It would be better to get the actual details on the failure by following the instructions output by gradle, i.e. running with `--stacktrace, --info, or --debug`.

Finally, to answer your question, yes, we are able to build this project on multiple independent systems, using the the exact code that is out on GitHub, both on developer workstations and also via CI, without any problems. Again, this suggests that your environment is not configured properly as required by the NDK, not a problem with the codebase itself.

**Comment by hannesa2** [comment]
> If I might ask, why did you switch the branch to your custom changes, instead of just checking out master as specified in the build steps that you originally suggested? 

Because without this change I run immediate in the compile issue (which it solves) and destroy the hope to make a step further. 
To have a clear way to demonstrate an issue (and have an evidence) I do all my work with a script, to prepare others an easy way to follow, that's way I need a branch

> The changes that you made on `FollowAdviceByNickParker` were only experimental, simply to try working around your specific reported problems. 

... I'm still investing why I run into issues

> Now that we have identified an environmental issue (i.e. that your system was missing a specific environmental variable required by the NDK), there is no reason to use the custom branch, particularly given that @developernotes verified your build steps were working on master.

That's not true
Even `export ANDROID_NDK_HOME=$ANDROID_NDK_ROOT` makes it not work

> Finally, to answer your question, yes, we are able to build this project on multiple independent systems, using the the exact code that is out on GitHub, both on developer workstations and also via CI, without any problems. Again, this suggests that your environment is not configured properly as required by the NDK, not a problem with the codebase itself.

CI ? Interesting. In repo there is no CI control yaml, or similar

**Comment by hannesa2** [comment]
I made it work with macOS ! But I'm not understanding exactly why

# macOS:

Testing on `master` master branch. 
A: Setting environment variables in shell doesn't work.
B: When I out varibles to ~/.bash_rc
```
  export ANDROID_NDK_ROOT=~/CipherbuildTest/android-ndk-r15c
  export ANDROID_NDK_HOME=$ANDROID_NDK_ROOT
  export PATH="/usr/local/opt/python@2/libexec/bin:$PATH":$ANDROID_NDK_ROOT
```
and do a `source ~/.bash_rc` it still not working.
C: But **after reboot it works**. No clue why ? Any ideas ?

# Linux

Very important:  
I've to do `git config --global core.autocrlf false` otherwise all scripts doesn't work

put values to `/etc/environment` 

```
export ANDROID_NDK_ROOT=/home/hannes/CipherbuildTest/android-ndk-r15c
export ANDROID_NDK_HOME=/home/hannes/CipherbuildTest/android-ndk-r15c
PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games":/home/hannes/CipherbuildTest/android-ndk-r15c
```
it shows me this
> /home/hannes/CipherbuildTest/android-ndk-r15c/build/gmsl/__gmsl:512: *** non-numeric second argument to `wordlist' function: ''.  Stop.

![image](https://user-images.githubusercontent.com/3314607/59875647-8f050780-93a1-11e9-84cd-99a9e2955839.png)

# general 

I want to make it work with Linux. This is a pre task to prepare a CI like Travis

**Comment by hannesa2** [comment]
I made it work on macOS and Linux. The main reason was Gradle Daemon.
Because I want to prepare a CI, I tried to prevent from settings values in `.profile`, `.bash_rc` and `.bash_profile` That's why I started build, to see how it fails, and after set values to make it work -this was my idea-
But nothing was working, till reboot !?

A simply `./gradlew --stop` solved it !
Explaination: first failed Gradle Daemon was started with unsetted environment, this daemon ignored all my attempt to set environment, because it was still started/running.

Now I figured out **no** `ANDROID_NDK_HOME` is needed, only the **PATH needs the NDK**
```
export ANDROID_NDK_ROOT=`pwd`/android-ndk-r15c
PATH=${PATH}:$ANDROID_NDK_ROOT
```
